e220/e220.go

```diff
@@ -329,34 +338,56 @@ func (d *Device) IsReady() bool {
 	return d.aux.Get()
 }
 
-// WriteMsg sends msg to the device corresponding to addr and ch
-// TODO: WriteMsg -> Write ??
-// TODO: Think of changing io.Writer
-func (d *Device) WriteMsg(addr uint16, ch uint8, msg []byte) (int, error) {
-	if len(msg) > (bufferSize - 3) {
-		return 0, fmt.Errorf("msg must be %dbytes or less: got=%d", bufferSize-3, len(msg))
-	}
-	// Wait for buffer is empty
-	for !d.IsReady() {
-	}
-	d.payload[0] = byte((addr & 0xFF00) >> 8)
-	d.payload[1] = byte((addr & 0x00FF) >> 0)
-	d.payload[2] = ch
-	for i := 0; i < len(msg); i++ {
-		d.payload[i+3] = msg[i]
+const (
+	TxMethodTransparent = iota
+	TxMethodFixed
+)
+
+// SetTxInfo sets information such as the destination address
+func (d *Device) SetTxInfo(addr uint16, ch uint8, txMethod int) error {
+	d.txAddr = addr
+	d.txCh = ch
+	if (txMethod != TxMethodTransparent) && (txMethod != TxMethodFixed) {
+		return fmt.Errorf("invalid Transmit Method: %d", d.txMethod)
 	}
-	return d.uart.Write(d.payload[:3+len(msg)])
+	d.txMethod = txMethod
 
+	return nil
 }
 
+// Write sends msg to the device corresponding to addr and ch
 func (d *Device) Write(p []byte) (n int, err error) {
-	if len(p) > (bufferSize) {
-		return 0, fmt.Errorf("msg must be %dbytes or less: got=%d", bufferSize, len(p))
+	var offset int
+	if d.txMethod == TxMethodTransparent {
+		offset = 0
+	} else if d.txMethod == TxMethodFixed {
+		d.payload[0] = byte((d.txAddr & 0xFF00) >> 8)
+		d.payload[1] = byte((d.txAddr & 0x00FF) >> 0)
+		d.payload[2] = d.txCh
+		offset = 3
+	} else {
+		return 0, fmt.Errorf("invalid transmit method: %d", d.txMethod)
+	}
+
+	if len(p) > (bufferSize - offset) {
+		return 0, fmt.Errorf("p must be %dbytes or less: got=%d", bufferSize-offset, len(p))
+	}
+	for i := 0; i < len(p); i++ {
+		d.payload[i+offset] = p[i]
 	}
```

Comment:
これは基本的に `copy()` を使うべき。
たぶん、こんな感じ。 (未検証

```suggestion
	copy(d.payload[offset:], p)
```

Reply:
dc167c0350d047a653ddd6c7e7331b39e76aa3ab にて対応しました。

1 年前くらいに Arudino Uno で copy が動かなかった記憶(勘違いかもしれないですが)があったので、念のため以下のコードが Arduino Uno で意図通り動作することを確認しました。

```go
package main

import "time"

func main() {
	a := []byte("Hello")
	b := make([]byte, 10)
	copy(b, a)
	for {
		println(string(b))
		time.Sleep(1000 * time.Millisecond)
	}
}
```